presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/provider/GSheetsDataProvider.java

```diff
@@ -0,0 +1,181 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.google.sheets.provider;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.sheets.v4.Sheets;
+import com.google.api.services.sheets.v4.SheetsScopes;
+import com.google.api.services.sheets.v4.model.ValueRange;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.airlift.log.Logger;
+import io.prestosql.plugin.google.sheets.SheetsConfig;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+public class GSheetsDataProvider
+        extends SheetsDataProvider
+{
+    private static final Logger log = Logger.get(GSheetsDataProvider.class);
+    private static final String APPLICATION_NAME = "presto google sheets integration";
+    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+    private static final List<String> SCOPES =
+            Collections.singletonList(SheetsScopes.SPREADSHEETS_READONLY);
+
+    // tableName <-> sheetId mapping
+    private final LoadingCache<String, String> tableSheetMap;
+
+    // sheetId <-> Row List mapping
+    private final LoadingCache<String, List<List<Object>>> sheetDataCache;
+
+    private Credential credential;
+    private NetHttpTransport httpTransport;
+    private Sheets sheetsService;
+
+    public GSheetsDataProvider(SheetsConfig config)
+    {
+        super(config);
+        try {
+            this.credential = getCredentials();
+            this.httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+            this.sheetsService = new Sheets.Builder(httpTransport, JSON_FACTORY, this.credential).setApplicationName(APPLICATION_NAME).build();
+        }
+        catch (Exception e) {
+            log.error(e, "Unable to load metadata mapping from %s", config.getMetadataSheetId());
+        }
+        this.tableSheetMap = CacheBuilder.newBuilder()
+            .expireAfterAccess(sheetsConfig.getSheetsDataExpireAfterAccessMinutes(), TimeUnit.MINUTES)
+            .expireAfterWrite(sheetsConfig.getSheetsDataExpireAfterWriteMinutes(), TimeUnit.MINUTES)
+            .build(
+                new CacheLoader<String, String>()
+                {
+                    public String load(String tableName)
+                    {
+                        try {
```

Comment:
extract this block as method. In general please follow the convention like in `io.prestosql.plugin.hive.metastore.CachingHiveMetastore`